feat: add cw721-non-transferable contract

eliseoci opened the pull request:
This PR includes a contract for implementing non-transferable (aka soulbound) NFTs.

eliseoci committed:
feat: add cw721-non-transferable contract

JakeHartnell commented on contracts/cw721-non-transferable/src/msg.rs:
```diff
@@ -0,0 +1,10 @@
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub admin: Option<Addr>,
+    pub cw721_instantiate_msg: Cw721InstantiateMsg,
```
May be better UX for frontend devs if you just inline the `Cw721InstantiateMsg` here? That way, could for existing NFT minting would _just work_ as the `admin` field is optional.

eliseoci replied:
makes sense! - comment addressed

larry0x commented on contracts/cw721-non-transferable/src/msg.rs:
```diff
@@ -0,0 +1,11 @@
+
+#[cw_serde]
+pub struct InstantiateMsg {
+    pub admin: Option<Addr>,
```
`admin` should be `Option<String>`. The instantiation function should validate the address, e.g.

```rust
// `msg.admin` is of `Option<String>` type  
let admin_addr: Addr = msg
    .admin
    .map(|s| deps.api.addr_validate(&s))
    .transpose()?;
```

0xekez replied:
if you're feeling really fancy, you can even get the value inside the option as a reference to avoid a move:

```rust
let admin = msg
    .admin  // Option<String>
    .as_deref()  // Option<&str>
    .map(|h| deps.api.addr_validate(h))  // Option<StdResult<Addr>>
    .transpose()?;  // StdResult<Option<Addr>>
```

eliseoci replied:
fixed it, ty both

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+                ExecuteMsg::Mint(msg) => {
+                    Cw721NonTransferableContract::default().mint(deps, env, info, msg)
+                }
+                _ => Err(ContractError::Unauthorized {}),
```
Personally I prefer more descriptive error messages than just "unauthorized", e.g. something like "only admin can transfer this token". However I understand this requires modifying the error type in `cw721-base` which is out-of-scope for this PR.

eliseoci replied:
+1 for more descriptive errors, but as you said, adding a new error type to the `cw721-base` is out-of-scope, will take it into account for a future PR

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+pub mod entry {
+    use super::*;
+    use cosmwasm_std::entry_point;
+    use cosmwasm_std::{Binary, Deps, DepsMut, Env, MessageInfo, Response, StdResult};
```
These two lines can be combined into a single statement

eliseoci replied:
fixed

larry0x commented on contracts/cw721-non-transferable/Cargo.toml:
```diff
@@ -0,0 +1,38 @@
+authors = [
+  "Eliseo Cohen <eliseoci@users.noreply.github.com>",
+]
+version = "0.1.0"
```
All other crates in this repo are at version `0.15.0`. Keep consistent?

eliseoci replied:
makes sense - changed it to `0.15.0`

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+    ) -> Result<Response, ContractError> {
+        cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        let config = Config { admin: msg.admin };
```
As mentioned in a separate comment, the admin address should be validated first.

eliseoci replied:
validation added

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+            env,
+            info,
+            cw721_base_instantiate_msg,
+        )?;
```
This will overwrite your cw2 version set above.

eliseoci replied:
fixed - good catch, thanks

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,85 @@
+
+    #[entry_point]
+    pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+        Cw721NonTransferableContract::default().query(deps, env, msg)
```
Imo there should be a query message for the admin address

eliseoci replied:
agree, done

larry0x reviewed (commented):
Three major issues I found:
- admin address is not validated
- cw2 contract version is overwritten
- there isn't a query method for the admin address

eliseoci commented:
Thanks for the great feedback @larry0x! I've addressed your comments. Lmk if it looks good to you.

On another topic, I've mistakenly opened the PR from my fork's main branch instead of the feature one. I can open a new PR and close this one to not have those annoying `merge pull request` commits in the log history, if you prefer.

larry0x commented:
> I can open a new PR and close this one to not have those annoying `merge pull request` commits

I would just squash-merge

larry0x commented on contracts/cw721-non-transferable/src/msg.rs:
```diff
@@ -0,0 +1,97 @@
+
+#[cw_serde]
+pub enum QueryMsg {
+    GetAdmin {},
```
I would just call this `Admin` instead of `GetAdmin`. All other query messages are noun phrases, while only this one is a verb phrase.

eliseoci replied:
Noted - just fixed it. I've squashed merge the previous commits as well.

larry0x commented on contracts/cw721-non-transferable/src/msg.rs:
```diff
@@ -0,0 +1,97 @@
+}
+
+#[cw_serde]
+pub struct GetAdminResponse {
```
Rename to `AdminResponse`?

eliseoci replied:
done

larry0x commented on contracts/cw721-non-transferable/src/msg.rs:
```diff
@@ -0,0 +1,97 @@
+
+#[cw_serde]
+pub struct GetAdminResponse {
+    pub admin: Option<Addr>,
```
Should be `Option<String>` in the response type as well

eliseoci replied:
done

larry0x commented on contracts/cw721-non-transferable/src/query.rs:
```diff
@@ -0,0 +1,10 @@
+use crate::{msg::GetAdminResponse, state::CONFIG};
+use cosmwasm_std::{Deps, Env, StdResult};
+
+pub fn admin(deps: Deps, _env: Env) -> StdResult<GetAdminResponse> {
```
`_env` can be removed if it’s not used in the function

eliseoci replied:
done, will take it into account from now on

eliseoci committed:
refactor: removed unused params + updated admin query response name

larry0x commented on contracts/cw721-non-transferable/src/lib.rs:
```diff
@@ -0,0 +1,98 @@
+        )?;
+
+        cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)
+            .map_err(ContractError::Std)?;
```
```suggestion
        cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
```

larry0x reviewed (commented):
One final comment. Everything else looks good to me

JakeHartnell committed:
Update contracts/cw721-non-transferable/src/lib.rs

Co-authored-by: Larry Engineer <26318510+larry0x@users.noreply.github.com>
